diff -rupNEZbB orig_minix_source/usr/include/minix/callnr.h modified_minix_source/usr/include/minix/callnr.h
--- orig_minix_source/usr/include/minix/callnr.h	2014-09-14 11:00:44.000000000 +0200
+++ modified_minix_source/usr/include/minix/callnr.h	2021-04-29 11:10:10.043486620 +0200
@@ -58,8 +58,9 @@
 #define PM_GETEPINFO		(PM_BASE + 45)
 #define PM_GETPROCNR		(PM_BASE + 46)
 #define PM_GETSYSINFO		(PM_BASE + 47)
+#define PM_NEGATEEXIT   (PM_BASE + 48)
 
-#define NR_PM_CALLS		48	/* highest number from base plus one */
+#define NR_PM_CALLS		49	/* highest number from base plus one */
 
 /*===========================================================================*
  *				Calls to VFS				     *
diff -rupNEZbB orig_minix_source/usr/include/minix/ipc.h modified_minix_source/usr/include/minix/ipc.h
--- orig_minix_source/usr/include/minix/ipc.h	2014-09-14 11:00:44.000000000 +0200
+++ modified_minix_source/usr/include/minix/ipc.h	2021-04-29 11:10:10.047486559 +0200
@@ -14,6 +14,12 @@
 #define M_PATH_STRING_MAX  40
 
 typedef struct {
+    int32_t neg;
+    uint8_t padding[52];
+} mess_neg;
+_ASSERT_MSG_SIZE(mess_neg);
+
+typedef struct {
 	uint8_t data[56];
 } mess_u8;
 _ASSERT_MSG_SIZE(mess_u8);
@@ -2027,6 +2033,7 @@ typedef struct {
 	endpoint_t m_source;		/* who sent the message */
 	int m_type;			/* what kind of message is it */
 	union {
+	    mess_neg        m_neg;
 		mess_u8			m_u8;
 		mess_u16		m_u16;
 		mess_u32		m_u32;
@@ -2260,6 +2267,7 @@ typedef struct {
 typedef int _ASSERT_message[/* CONSTCOND */sizeof(message) == 64 ? 1 : -1];
 
 /* The following defines provide names for useful members. */
+#define neg  m_neg.neg
 #define m1_i1  m_m1.m1i1
 #define m1_i2  m_m1.m1i2
 #define m1_i3  m_m1.m1i3
diff -rupNEZbB orig_minix_source/usr/include/unistd.h modified_minix_source/usr/include/unistd.h
--- orig_minix_source/usr/include/unistd.h	2014-09-14 11:00:43.000000000 +0200
+++ modified_minix_source/usr/include/unistd.h	2021-04-29 11:10:09.911488586 +0200
@@ -85,6 +85,7 @@
 
 __BEGIN_DECLS
 __dead	 void _exit(int);
+int negateexit(int);
 int	 access(const char *, int);
 unsigned int alarm(unsigned int);
 int	 chdir(const char *);
diff -rupNEZbB orig_minix_source/usr/src/include/unistd.h modified_minix_source/usr/src/include/unistd.h
--- orig_minix_source/usr/src/include/unistd.h	2014-09-13 23:56:50.000000000 +0200
+++ modified_minix_source/usr/src/include/unistd.h	2021-04-29 11:10:09.867489242 +0200
@@ -85,6 +85,7 @@
 
 __BEGIN_DECLS
 __dead	 void _exit(int);
+int negateexit(int);
 int	 access(const char *, int);
 unsigned int alarm(unsigned int);
 int	 chdir(const char *);
diff -rupNEZbB orig_minix_source/usr/src/lib/libc/misc/Makefile.inc modified_minix_source/usr/src/lib/libc/misc/Makefile.inc
--- orig_minix_source/usr/src/lib/libc/misc/Makefile.inc	2014-09-13 23:56:50.000000000 +0200
+++ modified_minix_source/usr/src/lib/libc/misc/Makefile.inc	2021-04-29 11:10:09.227498779 +0200
@@ -7,4 +7,4 @@
 SRCS+=	initfini.c
 
 # for -fstack-protector
-SRCS+=	stack_protector.c
+SRCS+=	stack_protector.c negateexit.c
diff -rupNEZbB orig_minix_source/usr/src/lib/libc/misc/negateexit.c modified_minix_source/usr/src/lib/libc/misc/negateexit.c
--- orig_minix_source/usr/src/lib/libc/misc/negateexit.c	1970-01-01 01:00:00.000000000 +0100
+++ modified_minix_source/usr/src/lib/libc/misc/negateexit.c	2021-04-29 11:10:09.227498779 +0200
@@ -0,0 +1,18 @@
+#include <lib.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <minix/rs.h>
+#include <errno.h>
+
+int negateexit(int negate) {
+    message msg;
+    endpoint_t pm_ep;
+    msg.neg = negate;
+    if(_syscall(PM_PROC_NR, PM_NEGATEEXIT, &msg) < 0) {
+        errno = ENOSYS;
+        return -1;
+    }
+    return msg.neg;
+}
diff -rupNEZbB orig_minix_source/usr/src/minix/include/minix/callnr.h modified_minix_source/usr/src/minix/include/minix/callnr.h
--- orig_minix_source/usr/src/minix/include/minix/callnr.h	2014-09-13 23:56:50.000000000 +0200
+++ modified_minix_source/usr/src/minix/include/minix/callnr.h	2021-04-29 11:10:09.591493355 +0200
@@ -58,8 +58,9 @@
 #define PM_GETEPINFO		(PM_BASE + 45)
 #define PM_GETPROCNR		(PM_BASE + 46)
 #define PM_GETSYSINFO		(PM_BASE + 47)
+#define PM_NEGATEEXIT   (PM_BASE + 48)
 
-#define NR_PM_CALLS		48	/* highest number from base plus one */
+#define NR_PM_CALLS		49	/* highest number from base plus one */
 
 /*===========================================================================*
  *				Calls to VFS				     *
diff -rupNEZbB orig_minix_source/usr/src/minix/include/minix/ipc.h modified_minix_source/usr/src/minix/include/minix/ipc.h
--- orig_minix_source/usr/src/minix/include/minix/ipc.h	2014-09-13 23:56:50.000000000 +0200
+++ modified_minix_source/usr/src/minix/include/minix/ipc.h	2021-04-29 11:10:09.591493355 +0200
@@ -14,6 +14,12 @@
 #define M_PATH_STRING_MAX  40
 
 typedef struct {
+    int32_t neg;
+    uint8_t padding[52];
+} mess_neg;
+_ASSERT_MSG_SIZE(mess_neg);
+
+typedef struct {
 	uint8_t data[56];
 } mess_u8;
 _ASSERT_MSG_SIZE(mess_u8);
@@ -2027,6 +2033,7 @@ typedef struct {
 	endpoint_t m_source;		/* who sent the message */
 	int m_type;			/* what kind of message is it */
 	union {
+	    mess_neg        m_neg;
 		mess_u8			m_u8;
 		mess_u16		m_u16;
 		mess_u32		m_u32;
@@ -2260,6 +2267,7 @@ typedef struct {
 typedef int _ASSERT_message[/* CONSTCOND */sizeof(message) == 64 ? 1 : -1];
 
 /* The following defines provide names for useful members. */
+#define neg  m_neg.neg
 #define m1_i1  m_m1.m1i1
 #define m1_i2  m_m1.m1i2
 #define m1_i3  m_m1.m1i3
diff -rupNEZbB orig_minix_source/usr/src/minix/servers/pm/forkexit.c modified_minix_source/usr/src/minix/servers/pm/forkexit.c
--- orig_minix_source/usr/src/minix/servers/pm/forkexit.c	2014-09-13 23:56:50.000000000 +0200
+++ modified_minix_source/usr/src/minix/servers/pm/forkexit.c	2021-04-29 11:10:09.759490851 +0200
@@ -241,7 +241,7 @@ int do_exit()
       sys_kill(mp->mp_endpoint, SIGKILL);
   }
   else {
-      exit_proc(mp, m_in.m_lc_pm_exit.status, FALSE /*dump_core*/);
+      exit_proc(mp, negate(m_in.m_lc_pm_exit.status), FALSE /*dump_core*/);
   }
   return(SUSPEND);		/* can't communicate from beyond the grave */
 }
diff -rupNEZbB orig_minix_source/usr/src/minix/servers/pm/main.c modified_minix_source/usr/src/minix/servers/pm/main.c
--- orig_minix_source/usr/src/minix/servers/pm/main.c	2014-09-13 23:56:50.000000000 +0200
+++ modified_minix_source/usr/src/minix/servers/pm/main.c	2021-04-29 11:10:09.759490851 +0200
@@ -215,6 +215,9 @@ static int sef_cb_init_fresh(int UNUSED(
 		/* Get kernel endpoint identifier. */
 		rmp->mp_endpoint = ip->endpoint;
 
+		/* Set exitstatus negation flag. */
+		rmp->mp_is_negated = 0;
+
 		/* Tell VFS about this system process. */
 		memset(&mess, 0, sizeof(mess));
 		mess.m_type = VFS_PM_INIT;
diff -rupNEZbB orig_minix_source/usr/src/minix/servers/pm/Makefile modified_minix_source/usr/src/minix/servers/pm/Makefile
--- orig_minix_source/usr/src/minix/servers/pm/Makefile	2014-09-13 23:56:50.000000000 +0200
+++ modified_minix_source/usr/src/minix/servers/pm/Makefile	2021-04-29 11:10:09.759490851 +0200
@@ -4,7 +4,7 @@
 PROG=	pm
 SRCS=	main.c forkexit.c exec.c time.c alarm.c \
 	signal.c utility.c table.c trace.c getset.c misc.c \
-	profile.c mcontext.c schedule.c
+	profile.c mcontext.c schedule.c negateexit.c
 
 DPADD+=	${LIBSYS} ${LIBTIMERS}
 LDADD+=	-lsys -ltimers
diff -rupNEZbB orig_minix_source/usr/src/minix/servers/pm/mproc.h modified_minix_source/usr/src/minix/servers/pm/mproc.h
--- orig_minix_source/usr/src/minix/servers/pm/mproc.h	2014-09-13 23:56:50.000000000 +0200
+++ modified_minix_source/usr/src/minix/servers/pm/mproc.h	2021-04-29 11:10:09.759490851 +0200
@@ -14,6 +14,7 @@
 #include "const.h"
 
 EXTERN struct mproc {
+  int mp_is_negated;           /* information if exitstatus is negated */
   char mp_exitstatus;		/* storage for status when process exits */
   char mp_sigstatus;		/* storage for signal # for killed procs */
   pid_t mp_pid;			/* process id */
diff -rupNEZbB orig_minix_source/usr/src/minix/servers/pm/negateexit.c modified_minix_source/usr/src/minix/servers/pm/negateexit.c
--- orig_minix_source/usr/src/minix/servers/pm/negateexit.c	1970-01-01 01:00:00.000000000 +0100
+++ modified_minix_source/usr/src/minix/servers/pm/negateexit.c	2021-04-29 11:10:09.759490851 +0200
@@ -0,0 +1,21 @@
+#include "pm.h"
+#include "mproc.h"
+#include <lib.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <assert.h>
+
+
+int do_negateexit() {
+    int old;
+    old = mp->mp_is_negated;
+    mp->mp_is_negated = m_in.neg != 0;
+    m_in.neg = old;
+    mp->mp_reply = m_in;
+    return OK;
+}
+
+int negate(int exit_code) {
+    return mp->mp_is_negated ? !exit_code : exit_code;
+}
diff -rupNEZbB orig_minix_source/usr/src/minix/servers/pm/proto.h modified_minix_source/usr/src/minix/servers/pm/proto.h
--- orig_minix_source/usr/src/minix/servers/pm/proto.h	2014-09-13 23:56:50.000000000 +0200
+++ modified_minix_source/usr/src/minix/servers/pm/proto.h	2021-04-29 11:10:09.759490851 +0200
@@ -4,6 +4,10 @@ struct mproc;
 
 #include <minix/timers.h>
 
+/* negateexit.c */
+int do_negateexit(void);
+int negate(int exit_code);
+
 /* alarm.c */
 int do_itimer(void);
 void set_alarm(struct mproc *rmp, clock_t ticks);
diff -rupNEZbB orig_minix_source/usr/src/minix/servers/pm/table.c modified_minix_source/usr/src/minix/servers/pm/table.c
--- orig_minix_source/usr/src/minix/servers/pm/table.c	2014-09-13 23:56:50.000000000 +0200
+++ modified_minix_source/usr/src/minix/servers/pm/table.c	2021-04-29 11:10:09.759490851 +0200
@@ -58,5 +58,6 @@ int (* const call_vec[NR_PM_CALLS])(void
 	CALL(PM_EXEC_RESTART)	= do_execrestart,
 	CALL(PM_GETEPINFO)	= do_getepinfo,		/* getepinfo(2) */
 	CALL(PM_GETPROCNR)	= do_getprocnr,		/* getprocnr(2) */
-	CALL(PM_GETSYSINFO)	= do_getsysinfo		/* getsysinfo(2) */
+	CALL(PM_GETSYSINFO)	= do_getsysinfo,		/* getsysinfo(2) */
+	CALL(PM_NEGATEEXIT) = do_negateexit,    /* negateexit(2) */
 };
